Thanks for the feedback!

I couldn’t reproduce the ICE before, but after enabling `[rust] debug-assertions = true` in `bootstrap.toml` (and `config.toml`) as you suggested and checking out the 1.79.0 tag, I was finally able to reproduce it locally.

Specifically, running the following code triggered the same ICE as before:

```rust
fn bar() -> impl Into<
    [u8; {
        let _ = for<'a, 'b> |x: &'a &'a Vec<&'b u32>, b: bool| -> &'a Vec<&'b u32> { *x };
    }],
> {
    [89]
}
```

**Version information**

comand: `.\build\x86_64-pc-windows-msvc\stage1\bin\rustc -vV`

```
rustc 1.79.0-dev
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-pc-windows-msvc
release: 1.79.0-dev
LLVM version: 18.1.7
```

<details>
<summary><b>Program output</b></summary>

```text
error[E0658]: `for<...>` binders for closures are experimental
  --> \rust\tests\ui\const-generics\generic_const_exprs\const-generics-closure.rs:3:17
   |
LL |         let _ = for<'a, 'b> |x: &'a &'a Vec<&'b u32>, b: bool| -> &'a Vec<&'b u32> { *x };
   |                 ^^^^^^^^^^^
   |
   = note: see issue #97362 <https://github.com/rust-lang/rust/issues/97362> for more information
   = help: add `#![feature(closure_lifetime_binder)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
   = help: consider removing `for<...>`

thread 'rustc' panicked at compiler\rustc_ast_lowering\src\item.rs:63:13:
assertion failed: matches!(owner, hir::MaybeOwner::Phantom)
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: <[rustc_span::symbol::Symbol] as rustc_data_structures::stable_hasher::HashStable<rustc_query_system::ich::hcx::StableHashingContext>>::hash_stable
   4: <rustc_ast_lowering::item::ItemLowerer>::lower_node
   5: rustc_ast_lowering::lower_to_hir
      [... omitted 2 frames ...]
   6: <&alloc::ffi::c_str::NulError as core::fmt::Debug>::fmt
      [... omitted 2 frames ...]
   7: rustc_middle::query::plumbing::query_get_at::<rustc_query_system::query::caches::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 8]>>>
   8: <rustc_middle::hir::map::Map>::get_module
   9: rustc_middle::hir::map::hir_crate_items
      [... omitted 2 frames ...]
  10: rustc_passes::hir_id_validator::check_crate
  11: rustc_interface::passes::create_global_ctxt
      [... omitted 2 frames ...]
  12: <rustc_middle::ty::context::GlobalCtxt>::enter::<rustc_driver_impl::run_compiler::{closure#0}::{closure#1}::{closure#3}, core::result::Result<(), rustc_span::ErrorGuaranteed>>
  13: RINvMs5_NtCs28QS2y6sdmQ_15rustc_interface7queriesNtNtB8_9interface8Compiler5enterNCNCNvCsdC0HzHJV9fq_17rustc_driver_impl12run_compiler0s_0INtNtCskla7WpyKoL0_4core6result6ResultINtNtB2h_6option6OptionNtB6_6LinkerENtCsjQRpMF1Y0aY_10rustc_span15ErrorGuarante
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.79.0-dev running on x86_64-pc-windows-msvc

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=\.cargo -Z ignore-directory-in-diagnostics-source-blocks=\rust\vendor -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0

query stack during panic:
#0 [hir_crate] getting the crate HIR
#1 [opt_hir_owner_nodes] getting HIR owner items in ``
#2 [hir_crate_items] getting HIR crate items
#3 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0658`.
```
</details>


I also cleaned up the test to remove the unrelated "cannot find function/value" errors while preserving the ICE.

let me know if anything else needs to be adjusted!

@petrochenkov

--- a/tests/ui/const-generics/generic_const_exprs/const-generics-closure.rs
+++ b/tests/ui/const-generics/generic_const_exprs/const-generics-closure.rs
@@ -1,0 +1,13 @@
+// Regression test for issue #127424
+
+fn bar() -> impl Into<
+    [u8; {
+        //~^ ERROR mismatched types [E0308]
+        let _ = for<'a, 'b> |x: &'a &'a Vec<&'b u32>, b: bool| -> &'a Vec<&'b u32> { *x };
+        //~^ ERROR `for<...>` binders for closures are experimental [E0658]
+    }],
+> {
+    [89]
+}
+
+fn main() {}

--- a/tests/ui/const-generics/generic_const_exprs/const-generics-closure.stderr
+++ b/tests/ui/const-generics/generic_const_exprs/const-generics-closure.stderr
@@ -1,0 +1,26 @@
+error[E0658]: `for<...>` binders for closures are experimental
+  --> $DIR/const-generics-closure.rs:6:17
+   |
+LL |         let _ = for<'a, 'b> |x: &'a &'a Vec<&'b u32>, b: bool| -> &'a Vec<&'b u32> { *x };
+   |                 ^^^^^^^^^^^
+   |
+   = note: see issue #97362 <https://github.com/rust-lang/rust/issues/97362> for more information
+   = help: add `#![feature(closure_lifetime_binder)]` to the crate attributes to enable
+   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
+   = help: consider removing `for<...>`
+
+error[E0308]: mismatched types
+  --> $DIR/const-generics-closure.rs:4:10
+   |
+LL |       [u8; {
+   |  __________^
+LL | |
+LL | |         let _ = for<'a, 'b> |x: &'a &'a Vec<&'b u32>, b: bool| -> &'a Vec<&'b u32> { *x };
+LL | |
+LL | |     }],
+   | |_____^ expected `usize`, found `()`
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0308, E0658.
+For more information about an error, try `rustc --explain E0308`.

